File: selenium-java/src/test/java/com/maelstromops/qa/pages/CustomerProfilePage.java
package com.maelstromops.qa.pages;

import com.maelstromops.qa.support.Config;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;

public class CustomerProfilePage extends BasePage {

  private final By root = By.cssSelector("[data-testid='customer.profile']");
  private final By name = By.id("name");
  private final By editButton = By.xpath("//button[normalize-space()='Edit']");
  private final By saveButton = By.xpath("//button[normalize-space()='Save']");
  private final By deleteButton = By.xpath("//button[normalize-space()='Delete']");

  public CustomerProfilePage(WebDriver driver) {
    super(driver);
  }

  public CustomerProfilePage open(String sysId) {
    driver.get(Config.baseUrl() + "/customer/browse/" + sysId + "/profile");
    visible(root);
    return this;
  }

  public boolean isLoaded() {
    return isVisible(root);
  }

  public boolean containsText(String text) {
    return wait.until(ExpectedConditions.textToBePresentInElementLocated(root, text));
  }

  /** Enter edit mode, change the name, and save (drops back to the read-only view). */
  public void rename(String newName) {
    click(editButton);
    type(name, newName);
    click(saveButton);
  }

  /** Click Delete and accept the native confirm dialog it raises. */
  public void deleteCustomer() {
    click(deleteButton);
    wait.until(ExpectedConditions.alertIsPresent()).accept();
  }
}

File: playwright-ts/src/pages/CustomerProfilePage.ts
import { expect, type Locator, type Page } from '@playwright/test';

/**
 * Page Object for a customer's profile, which has a read-only View that toggles
 * into Edit (Save/Cancel) and a Delete that confirms via a native dialog.
 */
export class CustomerProfilePage {
  readonly page: Page;
  readonly root: Locator;
  readonly editButton: Locator;
  readonly saveButton: Locator;
  readonly deleteButton: Locator;
  readonly nameInput: Locator;

  constructor(page: Page) {
    this.page = page;
    this.root = page.getByTestId('customer.profile');
    this.editButton = page.getByRole('button', { name: 'Edit' });
    this.saveButton = page.getByRole('button', { name: 'Save' });
    this.deleteButton = page.getByRole('button', { name: 'Delete' });
    this.nameInput = page.locator('#name');
  }

  async goto(sysId: string): Promise<void> {
    await this.page.goto(`/customer/browse/${sysId}/profile`);
    await expect(this.root).toBeVisible();
  }

  /** Enter edit mode, change the name, and save (drops back to read-only view). */
  async rename(newName: string): Promise<void> {
    await this.editButton.click();
    await this.nameInput.fill(newName);
    await this.saveButton.click();
  }

  /** Click Delete and accept the browser confirm dialog it raises. */
  async deleteCustomer(): Promise<void> {
    this.page.once('dialog', (dialog) => void dialog.accept());
    await this.deleteButton.click();
  }
}
